src/main/java/rueppellii/backend2/tribes/progression/service/ProgressionService.java

```diff
@@ -156,6 +159,22 @@ public void generateTroopUpgradeModel(Integer level, Kingdom kingdom) {
         }
     }
 
+    public List<Long> getTroopsForUpgradeIds(List<Troop> troopsForUpgrade) {
+        return troopsForUpgrade.stream().map(Troop::getId).collect(Collectors.toList());
+    }
+
+    public void troopUpgradeValidator(Kingdom kingdom, List<Long> troopsForUpgradeIds) throws UpgradeFailedException {
+        for(ProgressionModel troopProgression : kingdom.getKingdomsProgresses()) {
+            if (troopProgression.getType().equals("TROOP")) {
```

Comment:
We should talk about the refactor opportunity in here :)

Reply:
three ifs are now one :)